rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java

```diff
@@ -247,11 +259,26 @@ public void close() {
         try {
             flush();
             base.close();
+            if (cacheSnapshotHandler != null) {
+                cacheSnapshotHandler.save(committedCache);
+            }
             uncommittedCache.clear();
             committedCache.clear();
         }
         finally {
             this.lock.writeLock().unlock();
         }
     }
+
+    @Nonnull
+    private static Map<ByteArrayWrapper, byte[]> makeCommittedCache(int cacheSize,
```

Comment:
I think that maybe this method can be implemented as "initSnapshotHandler" or something like that, since its goal is to load the cache into the handler. I would leave the new constructor as this:

```java
public DataSourceWithCache(@Nonnull KeyValueDataSource base, int cacheSize,
                               @Nullable CacheSnapshotHandler cacheSnapshotHandler) {
    this.cacheSize = cacheSize;
    this.base = Objects.requireNonNull(base);
    this.uncommittedCache = new LinkedHashMap<>(cacheSize / 8, (float)0.75, false); 
    this.committedCache = Collections.synchronizedMap(new MaxSizeHashMap<>(cacheSize, true));
    this.cacheSnapshotHandler = cacheSnapshotHandler;
    initSnapshotHandler();
}
```

And then the other method would be as this:

```java
private void initSnapshotHandler() {
    if (this.cacheSnapshotHandler != null) {
            this.cacheSnapshotHandler.load(this.committedCache);
    }
}
```
Also we can get rid of this last method and apply that logic in the constructor. I think is a little bit more "direct" or easy to understand at plain sight.

Reply:
well, this method doesn't really initialise an instance of `SnapshotHandler`, but I'd say rather initialises `committedCache` map. In fact, `cacheSnapshotHandler` value is `nullable` (it is `null` for a few other data sources, like receipts and it is `null` when the feature is OFF (by default)) and class itself is stateless, so I guess it's not needed to be initialised.

I see `makeCommittedCache()` method as a factory that can create an instance of `Map<ByteArrayWrapper, byte[]>` and populate it, if needed.

Reply:
pipeline: run

Reply:
Yeah, that name `initSnapshotHandler` maybe is not the best to describe what's happening. It is more like a "configuration" or something like that. I mean: the `committedCache` will end up being initialized as `new MaxSizeHashMap<>(cacheSize, true);` no matter what, but the difference between the `cacheSnapshotHandler` being null or not only makes that object to load the already initialized `committedCache`. I think that the "flow" on the constructor would be a little bit clearer if we "configure" the `cacheSnapshotHandler` after all objects are assigned and so, that was the way I've seen it.